### DISCUSSION TITLE
How do I override pullquote fonts?

### DISCUSSION THREAD

**Discussion by chuckwondo** [sile-typesetter/sile#1691] [Q&A] [answered]
I'm sure my question is due to my lack of understanding some of the basics of SILE.

In the SILE manual, the section on pullquote includes the following:

> If you want to specify what font the pullquote environment should use, you can redefine the
`\pullquote:font` command. By default it will be the same as the surrounding document. The font
style used for the attribution line can likewise be set redefining `\pullquote:author-font` and the
font used for the quote marks can be set redefining `\pullquote:mark-font`.

I want to redefine `\pullquote:mark-font`, but I'm not sure I completely understand how to "redefine" a command.

Here's the gist of what I had at first:

```
\begin[author=Author]{pullquote}
A quote
\end{pullquote}
```

Since I don't have the 'Libertinus Serif' font installed, this gave me an error about that since that's the default font for `\pullquote:mark-font` (as I found by digging through the source code), so I want to specify a different font. (Of course, I could install that font, but I don't yet know how to install new fonts on macOS, so that's a separate matter.)

I finally found an example in the section about the `folio` package, so I got this to work with a font I already have installed (Palatino):

```
\define[command=pullquote:mark-font]{\font[family=Palatino]{\process}}

\begin[author=Author]{pullquote}
A quote
\end{pullquote}
```

So my questions are:

1. Is this the correct approach to "redefine" a command?
2. How do I prevent the attribution line from breaking from the quote? It just so happened that the pull quote ended up at the bottom of a page, but the attribution line didn't fit, so it was placed at the top of the next page. How do I prevent that from happening?
3. Is there another command I can use for inserting quotes that doesn't create a "pull" quote? I just want a quote that's more like the surrounding text, but just has some spacing above and below, some left indentation, and has an attribution line, not something so "dramatic" as a pull quote.
4. How do I remove the slight indent on the first line of the quote? In the SILE book, the rendered pull quote does not show an indent on the first line of the quote, but when I use it, I see a small indent, perhaps about the width of a character. How do I eliminate that, so that all lines of the quote are left aligned?

**Comment by Omikhleia** [comment, marked as answer]
1. This is one of the correct ways (using `\define`, i.e. a SILE macro), yes. (Another way would be to use Lua, but for simple things like this, macros are ok)
   ... but _in that very case_, the command hook expects a _global_ font to be set, so that would have to be, unless mistaken:
   ```
   \define[command=pullquote:mark-font]{\font[family=Palatino]}
   ```
   I.e. without the argument containing a `\process`. (That's one of the reason I don't like command hook redefinition, personally... and use other strategies in my own packages. There are too many possibilities for hooks to implement things, so "it depends".... But anyhow, as **pullquote** goes so far, well, that's it...)
1. There's currently no way with SILE macros to avoid a page break between the pullquote text and author. One could do it in Lua, and it could be a nice feature indeed. I just think this wasn't foreseen, heh. Or not use the author option and do the encoding all in-line... Something such the following might work:
   ```
   \begin{pullquote}Some quote
   \novbreak
   \raggedleft{— My author}
   \end{pullquote}
   ```
   (The trick is the `\novbreak` that tells SILE to avoid a page break here... This is what is missing from the Lua implementation.)
   That's kind of a workaround, though :smile_cat: 
1. No, no other good (block)quote in the standard distribution.[^1]
1. I am not quite sure what you mean by "the slight indent". If this is the offset I marked in red below...
   ![image](https://user-images.githubusercontent.com/18075640/214130719-9c1c923d-f426-409d-9898-3024b5914130.png)
   ... Then it's because of the way spaces are ignored or not in commands and environments (a matter of long debates).

   If so, in brief, you have an "extra" trailing space... Replace:
   ```
   \begin[author=Author]{pullquote}
   Some quote....
   \end{pullquote}
   ```
    with:
   ```
   \begin[author=Author]{pullquote}Some quote....\end{pullquote}
   ```
   or
   ```
   \begin[author=Author]{pullquote}%
   Some quote....
   \end{pullquote}
   ```

   or even `\pullquote[author=Author]{Some quote....}` (since commands and environments are actually kind of the same thing, that is just (theoretically) syntactic sugar)

[^1]: People may start to frown seeing me repeating the same things... But [my own classes](https://github.com/Omikhleia/resilient.sile) include a **blockquote** environment that can be nested, has (some) configurable settings via styles, for the amount of spacing, etc. I am still lagging behind regarding the work on upgrading these classes and packages, though, and the documentation is not available as a PDF yet.

**Reply by chuckwondo** [reply]
Fantastic! Thanks for your prompt and thorough replies.

**Reply by alerque** [reply]
I think @Omikhleia answered most of this question well, but I'd like to throw out that the spacing answer might be only half right. He's right about how a space got in there, but there is also passable *another bug* besides the space-after-command issue lurking here. I say this because I just ran into a similar case this weekend. Here is my output before some space hackery showing what I think is a different bug than the process spaces bailiwick.

![image](https://user-images.githubusercontent.com/173595/214318068-7d7c2abb-7097-423f-abba-bcc748ca39df.png)


Note question 11 starts out with a leading space that I wasn't expecting. I'm typesetting the numbers inside a rebox with a negative kern and rebox width set to the same value. There is no space in the SIL source between commands here, so that's not the issue. The only space being added was being added by me because I output the number with `SILE.typesetter:typeset(numer .. ". ")`. This was intended to fix the issue of copy/paste from the PDF not having the expected content.

I expected the `SILE.call("kern" { width = numwidth:negate() }); SILE.call("rebox", { width = numwidth })` to take care of this. And most of the time it is fine. But then in question 11 I ran into a line that was wrapped and suddenly instead of having a `document.rskip` of *hfill* we now have one of *0 plus 6psc*. Now suddenly the space inside the hanging indentation is getting stretched along with spaces in the rest of the line and it actually **breaks out of the rebox**! Oops.

We don't have a mechanism that would accommodate this use case. I can't remember if the *pullquote* package does something similar, but it might.

**Reply by alerque** [reply]
I looked at the *pullquote* package source and don't see any spaces leaking in in the way my example shows, so I think my case is a different box of shenanigans.
